File: service/src/main/java/no/nav/melosys/service/kontroll/feature/ufm/UfmKontrollService.java
package no.nav.melosys.service.kontroll.feature.ufm;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

import io.micrometer.core.instrument.Metrics;
import no.finn.unleash.Unleash;
import no.nav.melosys.domain.Behandling;
import no.nav.melosys.domain.Behandlingsresultat;
import no.nav.melosys.domain.Kontrollresultat;
import no.nav.melosys.domain.dokument.sed.SedDokument;
import no.nav.melosys.domain.eessi.SedType;
import no.nav.melosys.domain.kodeverk.begrunnelser.Kontroll_begrunnelser;
import no.nav.melosys.repository.KontrollresultatRepository;
import no.nav.melosys.service.behandling.BehandlingService;
import no.nav.melosys.service.behandling.BehandlingsresultatService;
import no.nav.melosys.service.behandlingsgrunnlag.BehandlingsgrunnlagService;
import no.nav.melosys.service.kontroll.feature.ufm.data.UfmKontrollData;
import no.nav.melosys.service.kontroll.feature.ufm.kontroll.UfmKontrollsett;
import no.nav.melosys.service.kontroll.regler.PeriodeRegler;
import no.nav.melosys.service.persondata.PersondataFasade;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.context.annotation.Primary;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import static no.nav.melosys.metrics.MetrikkerNavn.TAG_BEGRUNNELSE;
import static no.nav.melosys.metrics.MetrikkerNavn.UNNTAKSPERIODE_KONTROLL_TREFF;

@Service
@Primary
public class UfmKontrollService {
    private static final Logger log = LoggerFactory.getLogger(UfmKontrollService.class);

    static {
        Arrays.stream(Kontroll_begrunnelser.values())
            .forEach(b -> Metrics.counter(UNNTAKSPERIODE_KONTROLL_TREFF, TAG_BEGRUNNELSE, b.getKode()));
    }

    private final KontrollresultatRepository kontrollresultatRepository;
    private final BehandlingsresultatService behandlingsresultatService;
    private final BehandlingsgrunnlagService behandlingsgrunnlagService;
    private final BehandlingService behandlingService;
    private final PersondataFasade persondataFasade;
    private final Unleash unleash;

    public UfmKontrollService(KontrollresultatRepository kontrollresultatRepository,
                              BehandlingsresultatService behandlingsresultatService,
                              BehandlingsgrunnlagService behandlingsgrunnlagService, BehandlingService behandlingService,
                              PersondataFasade persondataFasade,
                              Unleash unleash) {
        this.kontrollresultatRepository = kontrollresultatRepository;
        this.behandlingsresultatService = behandlingsresultatService;
        this.behandlingsgrunnlagService = behandlingsgrunnlagService;
        this.behandlingService = behandlingService;
        this.persondataFasade = persondataFasade;
        this.unleash = unleash;
    }

    @Transactional
    public void utførKontrollerOgRegistrerFeil(long behandlingId) {
        Behandling behandling = behandlingService.hentBehandlingMedSaksopplysninger(behandlingId);
        List<Kontroll_begrunnelser> registrerteTreff = utførKontroller(behandling);

        log.info("Treff ved validering av periode for behandling {}: {}", behandlingId, registrerteTreff);
        lagreKontrollresultater(behandlingId, registrerteTreff);
    }

    List<Kontroll_begrunnelser> utførKontroller(Behandling behandling) {
        var sedDokument = behandling.hentSedDokument();
        if (harFeilIPeriode(sedDokument)) {
            return Collections.singletonList(Kontroll_begrunnelser.FEIL_I_PERIODEN);
        }
        var ufmKontrollData = lagUfmKontrollData(behandling, sedDokument);
        var sedType = sedDokument.getSedType();
        return utførKontroller(ufmKontrollData, sedType);
    }

    private UfmKontrollData lagUfmKontrollData(Behandling behandling, SedDokument sedDokument) {
        var persondata = persondataFasade.hentPerson(behandling.getFagsak().hentBrukersAktørID());
        var medlemskapDokument = behandling.hentMedlemskapDokument();
        var inntektDokument = behandling.hentInntektDokument();
        var utbetalingDokument = behandling.finnUtbetalingDokument().orElse(null);
        var optionalBehandlingsgrunnlagData = behandlingsgrunnlagService.finnBehandlingsgrunnlagdata(behandling.getId());
        return new UfmKontrollData(sedDokument, persondata, medlemskapDokument, inntektDokument, utbetalingDokument, optionalBehandlingsgrunnlagData);
    }

    private List<Kontroll_begrunnelser> utførKontroller(UfmKontrollData kontrollData, SedType sedType) {
        return UfmKontrollsett.hentRegelsettForSedType(sedType, unleash).stream()
            .map(f -> f.apply(kontrollData))
            .filter(Objects::nonNull)
            .peek(this::registrerMetrikk) //NOSONAR
            .toList();
    }

    private void lagreKontrollresultater(Long behandlingID, List<Kontroll_begrunnelser> kontrollBegrunnelser) {
        Behandlingsresultat behandlingsresultat = behandlingsresultatService.hentBehandlingsresultat(behandlingID);

        kontrollresultatRepository.deleteByBehandlingsresultat(behandlingsresultat);
        kontrollresultatRepository.flush();

        List<Kontrollresultat> kontrollresultater = kontrollBegrunnelser.stream()
            .map(kontrollBegrunnelse -> lagKontrollresultat(behandlingsresultat, kontrollBegrunnelse))
            .toList();

        kontrollresultatRepository.saveAll(kontrollresultater);
    }

    private Kontrollresultat lagKontrollresultat(Behandlingsresultat behandlingsresultat, Kontroll_begrunnelser kontrollBegrunnelse) {
        Kontrollresultat kontrollresultat = new Kontrollresultat();
        kontrollresultat.setBegrunnelse(kontrollBegrunnelse);
        kontrollresultat.setBehandlingsresultat(behandlingsresultat);

        return kontrollresultat;
    }

    private void registrerMetrikk(Kontroll_begrunnelser unntak_periode_begrunnelse) {
        Metrics.counter(UNNTAKSPERIODE_KONTROLL_TREFF, TAG_BEGRUNNELSE, unntak_periode_begrunnelse.getKode()).increment();
    }

    private boolean harFeilIPeriode(SedDokument sedDokument) {
        return PeriodeRegler.feilIPeriode(
            sedDokument.getLovvalgsperiode().getFom(),
            sedDokument.getLovvalgsperiode().getTom());
    }
}

File: service/src/main/kotlin/no/nav/melosys/service/kontroll/feature/ufm/UfmKontrollService.kt
package no.nav.melosys.service.kontroll.feature.ufm

import io.micrometer.core.instrument.Metrics
import jakarta.annotation.PostConstruct
import mu.KotlinLogging
import no.nav.melosys.domain.Behandling
import no.nav.melosys.domain.Behandlingsresultat
import no.nav.melosys.domain.Kontrollresultat
import no.nav.melosys.domain.dokument.person.PersonhistorikkDokument
import no.nav.melosys.domain.dokument.sed.SedDokument
import no.nav.melosys.domain.eessi.SedType
import no.nav.melosys.domain.kodeverk.begrunnelser.Kontroll_begrunnelser
import no.nav.melosys.metrics.MetrikkerNavn
import no.nav.melosys.repository.KontrollresultatRepository
import no.nav.melosys.service.behandling.BehandlingService
import no.nav.melosys.service.behandling.BehandlingsresultatService
import no.nav.melosys.service.kontroll.feature.ufm.data.UfmKontrollData
import no.nav.melosys.service.kontroll.feature.ufm.kontroll.UfmKontrollsett
import no.nav.melosys.service.kontroll.regler.PeriodeRegler
import no.nav.melosys.service.mottatteopplysninger.MottatteOpplysningerService
import no.nav.melosys.service.persondata.PersondataFasade
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

private val log = KotlinLogging.logger { }

@Service
class UfmKontrollService(
    private val kontrollresultatRepository: KontrollresultatRepository,
    private val behandlingsresultatService: BehandlingsresultatService,
    private val mottatteOpplysningerService: MottatteOpplysningerService,
    private val behandlingService: BehandlingService,
    private val persondataFasade: PersondataFasade,
) {

    @PostConstruct
    private fun registerKontrollBegrunnelser() {
        Kontroll_begrunnelser.values().forEach { begrunnelse ->
            Metrics.counter(
                MetrikkerNavn.UNNTAKSPERIODE_KONTROLL_TREFF,
                MetrikkerNavn.TAG_BEGRUNNELSE,
                begrunnelse.kode
            )
        }
    }

    @Transactional
    fun utførKontrollerOgRegistrerFeil(behandlingId: Long) {
        val behandling = behandlingService.hentBehandlingMedSaksopplysninger(behandlingId)
        val registrerteTreff = utførKontroller(behandling)

        log.info("Treff ved validering av periode for behandling {}: {}", behandlingId, registrerteTreff)
        lagreKontrollresultater(behandlingId, registrerteTreff)
    }

    fun utførKontroller(behandling: Behandling): List<Kontroll_begrunnelser> {
        val sedDokument = behandling.hentSedDokument()

        if (harFeilIPeriode(sedDokument)) {
            return listOf(Kontroll_begrunnelser.FEIL_I_PERIODEN)
        }
        val sedType = sedDokument.sedType
        val saksopplysninger = behandling.saksopplysninger
        val personhistorikkDokumenter = saksopplysninger
            .filter { it.dokument is PersonhistorikkDokument }
            .map { it.dokument as PersonhistorikkDokument }

        val ufmKontrollData = lagUfmKontrollData(behandling, personhistorikkDokumenter, sedDokument)

        return utførKontroller(ufmKontrollData, sedType)
    }

    private fun lagUfmKontrollData(
        behandling: Behandling,
        personhistorikkDokumenter: List<PersonhistorikkDokument?>,
        sedDokument: SedDokument
    ): UfmKontrollData {
        val persondata = persondataFasade.hentPerson(behandling.fagsak.hentBrukersAktørID())
        val persondataMedHistorikk = persondataFasade.hentPersonMedHistorikk(behandling.fagsak.hentBrukersAktørID())
        val medlemskapDokument = behandling.hentMedlemskapDokument()
        val inntektDokument = behandling.hentInntektDokument()
        val utbetalingDokument = behandling.finnUtbetalingDokument().orElse(null)
        val optionalMottatteOpplysningerData = mottatteOpplysningerService.finnMottatteOpplysningerData(behandling.id)
        return UfmKontrollData(
            sedDokument,
            persondata,
            medlemskapDokument,
            inntektDokument,
            utbetalingDokument,
            optionalMottatteOpplysningerData,
            personhistorikkDokumenter,
            Optional.of(persondataMedHistorikk)
        )
    }

    private fun utførKontroller(
        kontrollData: UfmKontrollData,
        sedType: SedType
    ): List<Kontroll_begrunnelser> = UfmKontrollsett.hentRegelsettForSedType(sedType)
        .mapNotNull { it(kontrollData) }
        .onEach { registrerMetrikk(it) }

    private fun lagreKontrollresultater(behandlingID: Long, kontrollBegrunnelser: List<Kontroll_begrunnelser>) {
        val behandlingsresultat = behandlingsresultatService.hentBehandlingsresultat(behandlingID)

        kontrollresultatRepository.deleteByBehandlingsresultat(behandlingsresultat)
        kontrollresultatRepository.flush()

        val kontrollresultater = kontrollBegrunnelser.stream()
            .map { kontrollBegrunnelse: Kontroll_begrunnelser ->
                lagKontrollresultat(
                    behandlingsresultat,
                    kontrollBegrunnelse
                )
            }.toList()

        kontrollresultatRepository.saveAll(kontrollresultater)
    }

    private fun lagKontrollresultat(behandlingsresultat: Behandlingsresultat, kontrollBegrunnelse: Kontroll_begrunnelser): Kontrollresultat {
        val kontrollresultat = Kontrollresultat()
        kontrollresultat.begrunnelse = kontrollBegrunnelse
        kontrollresultat.behandlingsresultat = behandlingsresultat

        return kontrollresultat
    }

    private fun registrerMetrikk(unntakPeriodeBegrunnelse: Kontroll_begrunnelser) {
        Metrics.counter(
            MetrikkerNavn.UNNTAKSPERIODE_KONTROLL_TREFF,
            MetrikkerNavn.TAG_BEGRUNNELSE, unntakPeriodeBegrunnelse.kode
        ).increment()
    }

    private fun harFeilIPeriode(sedDokument: SedDokument): Boolean = PeriodeRegler.feilIPeriode(
        sedDokument.lovvalgsperiode.fom,
        sedDokument.lovvalgsperiode.tom
    )
}
